File: src/main/java/org/cerion/stocklist/overlays/OverlayBase.java
package org.cerion.stocklist.overlays;


import org.cerion.stocklist.PriceList;
import org.cerion.stocklist.arrays.FloatArray;
import org.cerion.stocklist.arrays.ValueArray;
import org.cerion.stocklist.functions.types.IFunctionEnum;
import org.cerion.stocklist.functions.ISimpleOverlay;

import java.lang.reflect.Method;

abstract class OverlayBase<T extends ValueArray> extends PriceOverlayBase implements ISimpleOverlay {

    OverlayBase(IFunctionEnum id, Number ...params) {
        super(id,params);
    }

    @Override
    public T eval(PriceList list) {
        return eval(list.getClose());
    }

    @Override
    public abstract T eval(FloatArray arr);

    @Override
    @SuppressWarnings("unchecked")
    public Class<? extends ValueArray> getResultType() {
        try {
            Method evalMethod = getClass().getMethod("eval", FloatArray.class);
            return (Class<? extends ValueArray>)evalMethod.getReturnType();
        }
        catch (Exception e){
            throw new RuntimeException(e.getMessage());
        }
    }
}

File: src/main/java/org/cerion/stocklist/overlays/OverlayBase.kt
package org.cerion.stocklist.overlays


import org.cerion.stocklist.PriceList
import org.cerion.stocklist.arrays.FloatArray
import org.cerion.stocklist.arrays.ValueArray
import org.cerion.stocklist.functions.ISimpleOverlay
import org.cerion.stocklist.functions.types.IFunctionEnum

abstract class OverlayBase<T : ValueArray> internal constructor(id: IFunctionEnum, vararg params: Number) : PriceOverlayBase(id, *params), ISimpleOverlay {

    override fun eval(list: PriceList): T {
        return eval(list.close)
    }

    abstract override fun eval(arr: FloatArray): T

    override fun getResultType(): Class<out ValueArray> {
        try {
            val evalMethod = javaClass.getMethod("eval", FloatArray::class.java)
            return evalMethod.returnType as Class<out ValueArray>
        } catch (e: Exception) {
            throw RuntimeException(e.message)
        }
    }
}
